src/Halogen/DatePicker/Component/Interval.purs

```diff
@@ -0,0 +1,235 @@
+
+renderDuration :: ∀ m. DurationF.Format -> HTML m
+renderDuration fmt = HH.slot' cpDuration unit (Duration.picker fmt) unit (HE.input $ Update <<< Left)
+
+renderDateTime :: ∀ m. DateTimeF.Format -> Boolean -> HTML m
+renderDateTime fmt idx = HH.slot' cpDateTime idx (DateTime.picker fmt) unit (HE.input $ Update <<< Right <<< (Tuple idx))
+
+
+-- [1] - this case will not happen as interval will not be `Just Right`
+--       if any of it's child is `Nothing` so return nonsence value
+evalInterval ∷ ∀ m . IntervalQuery ~> DSL m
+evalInterval (Update msg next) = do
+  s <- H.get
+  nextInterval <- map (steper s.interval) $ case s.interval of
+    Nothing  -> do
+      interval <- buildInterval
+      case interval of
+        Left (Tuple false _) -> resetChildErrorBasedOnMessage msg
+        _ -> pure unit
+      pure interval
+    Just (Left err) -> buildInterval
+    Just (Right interval) -> pure $ lmap (Tuple false) $ case msg of
+      Left (NotifyChange newDuration) -> case newDuration of
+        Just (Left x) -> Left $ bimap (const $ Just x) (const Nothing) s.format
+        Nothing -> Left $ bimap (const Nothing) (const Nothing) s.format -- [1]
+        Just (Right duration) -> Right $ lmap (const duration) interval
+      Right (Tuple idx (NotifyChange newDateTime)) -> case newDateTime of
+        Just (Left x) -> Left $ bimap (const Nothing) (const $ Just x) s.format
+        Nothing -> Left $ bimap (const Nothing) (const Nothing) s.format -- [1]
+        Just (Right dateTime) -> Right $ case interval of
+          StartEnd a b -> case idx of
+            true -> StartEnd dateTime b
+            false -> StartEnd a dateTime
+          DurationEnd d a -> DurationEnd d dateTime
+          StartDuration a d -> StartDuration dateTime d
+          JustDuration d -> JustDuration d
+  H.modify _{ interval = nextInterval }
+  when (nextInterval /= s.interval) $ H.raise (NotifyChange nextInterval)
+  pure next
+
+
+buildInterval :: ∀ m. DSL m (Either (Tuple Boolean IntervalError) IsoInterval)
+buildInterval = do
+  {format} <- H.get
+  vals <- collectValues format
+  pure $ lmap addForce $ unVals vals
+
+addForce :: IntervalError -> Tuple Boolean IntervalError
+addForce err = case err of
+  StartEnd Nothing Nothing -> Tuple false err
+  DurationEnd Nothing Nothing -> Tuple false err
+  StartDuration Nothing Nothing -> Tuple false err
+  JustDuration Nothing -> Tuple false err
+  _ -> Tuple true err
+
+unVals :: Interval Duration.Input DateTime.Input -> Either IntervalError IsoInterval
+unVals vals = case bimap maybeLeft maybeLeft vals of
+  StartEnd (Right dtStart) (Right dtEnd) -> Right $ StartEnd dtStart dtEnd
+  DurationEnd (Right dur) (Right dt) -> Right $ DurationEnd dur dt
+  StartDuration (Right dt) (Right dur) -> Right $ StartDuration dt dur
+  JustDuration (Right dur) -> Right $ JustDuration dur
+  interval -> Left $ bimap toError toError interval
+
+toError :: ∀ e a. Either (Maybe e) a -> Maybe e
+toError = either id (const Nothing)
```

Comment:
This will work for any `Plus` or `Monoid`
```purescript
leftOrEmpty :: forall e m a. Plus m => Either (m a) a -> m a
leftOrEmpty = either id $ const empty
```

```purescript
leftOrEmpty:: forall m a b. Monoid m => Bifoldable b => b m a -> m 
leftOrEmpty = bifoldMap id $ const mempty
```
(Isn't blocker, just note)